File: swift/Sources/AOC2022/Day02.swift
private enum Hand {
    case rock
    case paper
    case scissors
}

extension Hand {
    static let beatsMap: [Hand: Hand] = [.rock: .scissors, .paper: .rock, .scissors: .paper]

    static let losesMap: [Hand: Hand] = [.scissors: .rock, .rock: .paper, .paper: .scissors]

    init(_ symbol: String) throws {
        switch symbol {
        case "A", "X": self = .rock
        case "B", "Y": self = .paper
        case "C", "Z": self = .scissors
        default: throw InputError.parseError("Can't parse \(symbol)")
        }
    }

    func beats() -> Hand {
        return Hand.beatsMap[self]!
    }

    func loses() -> Hand {
        return Hand.losesMap[self]!
    }

    func value() -> Int {
        switch self {
        case .rock:
            return 1
        case .paper:
            return 2
        case .scissors:
            return 3
        }
    }
}

enum Day02 {
    static func solve1(_ input: String) throws -> Int {
        let lines = input.components(separatedBy: "\n")
        return try lines
            .map { l in try score1(l) }
            .reduce(0, +)
    }

    static func score1(_ line: String) throws -> Int {
        let parts = line.components(separatedBy: " ")
        let theirs = try Hand(parts[0])
        let mine = try Hand(parts[1])

        if mine == theirs {
            return mine.value() + 3
        } else if mine.beats() == theirs {
            return mine.value() + 6
        } else {
            return mine.value()
        }
    }

    static func solve2(_ input: String) throws -> Int {
        let lines = input.components(separatedBy: "\n")
        return try lines
            .map { l in try score2(l) }
            .reduce(0, +)
    }

    static func score2(_ line: String) throws -> Int {
        let parts = line.components(separatedBy: " ")
        let theirs = try Hand(parts[0])

        // X means you need to lose, Y means you need to end the round in a draw, and Z means you need to win
        let symbol = parts[1]
        switch symbol {
        case "X":
            return theirs.beats().value()
        case "Y":
            return theirs.value() + 3
        case "Z":
            return theirs.loses().value() + 6
        default: throw InputError.parseError("Can't parse \(symbol)")
        }
    }
}

File: java/src/test/java/aoc2022/Day02.java
package aoc2022;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;


enum Hand {
    ROCK(1),
    PAPER(2),
    SCISSORS(3);

    final int value;

    Hand(int value) {
        this.value = value;
    }
}

// A for Rock, B for Paper, and C for Scissors
public class Day02 {

    public static long solve1(String input) {
        return input.lines()
                .mapToLong(Day02::score1)
                .sum();
    }

    public static long score1(String game) {
        final String[] s = game.split(" ");
        var theirs = parse(s[0]);
        var mine = parse(s[1]);

        if ((mine == Hand.ROCK && theirs == Hand.SCISSORS) ||
                (mine == Hand.PAPER && theirs == Hand.ROCK) ||
                (mine == Hand.SCISSORS && theirs == Hand.PAPER)) {
            return mine.value + 6;
        } else if (mine == theirs) {
            return mine.value + 3;
        } else {
            return mine.value;
        }
    }

    public static Hand parse(String s) {
        return switch (s) {
            case "A", "X" -> Hand.ROCK;
            case "B", "Y" -> Hand.PAPER;
            case "C", "Z" -> Hand.SCISSORS;
            default -> throw new IllegalStateException("Unknown input: " + s);
        };
    }

    public static long solve2(String input) {
        return input.lines()
                .mapToLong(Day02::score2)
                .sum();
    }

    public static long score2(String game) {
        final String[] s = game.split(" ");
        var theirs = parse(s[0]);
        var mine = s[1];

        // X means you need to lose, Y means you need to end the round in a draw, and Z means you need to win
        switch (mine) {
            case "X" -> {
                switch (theirs) {
                    case ROCK -> {
                        return Hand.SCISSORS.value;
                    }
                    case PAPER -> {
                        return Hand.ROCK.value;
                    }
                    case SCISSORS -> {
                        return Hand.PAPER.value;
                    }
                }
            }
            case "Y" -> {
                return theirs.value + 3;
            }
            case "Z" -> {
                switch (theirs) {
                    case ROCK -> {
                        return Hand.PAPER.value + 6;
                    }
                    case PAPER -> {
                        return Hand.SCISSORS.value + 6;
                    }
                    case SCISSORS -> {
                        return Hand.ROCK.value + 6;
                    }
                }
            }
        }
        return 0;
    }

    public static void main(String[] args) throws Exception {
        var input = Resources.readString(Resources.class.getResource("/day02.txt"));
        System.out.println(solve1(input));
        System.out.println(solve2(input));
    }

    @Test
    void example() {
        final String input = """
                A Y
                B X
                C Z
                """;
        assertEquals(15, solve1(input));
        assertEquals(12, solve2(input));
    }
}
